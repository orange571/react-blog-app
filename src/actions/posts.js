import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment'

// ADD_POST
export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      isReadable = false,
      author = '',
    } = postData;
    const post = { title, body, createdAt, isReadable, author, uid };

    return database.ref(`posts`).push(post).then((ref) => {
      const postId = ref.key;
      database.ref(`users/${uid}/posts`).update({[postId] : isReadable}).then(()=>{
        if(isReadable){
          database.ref(`isPublic`).push({id: postId})
        };
      });
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

// REMOVE_POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).remove().then(() => {
      //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
          const wasPublic = snapshot.val();
          if (!!wasPublic){
            const publicPostId = Object.keys(wasPublic)[0];
            database.ref(`isPublic/${publicPostId}`).remove().then(() => {
              //REMOVE FROM USER List
              database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
              .catch(error => console.log(error));
            }).catch(error => console.log(error));
          } else {
            //REMOVE FROM USER List
            database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
            .catch(error => console.log(error));
          }
      });
      dispatch(removePost({ id }));
    });
  };
};

// EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).update(updates).then(() => {
      //ADD TO PUBLIC LIST IF WAS NOT THERE ALREADY
      if(updates.isReadable){
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            if (!wasPublic){
              database.ref(`isPublic`).push({id: id})
            }
        });
        //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      } else {
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            const publicPostId = Object.keys(wasPublic)[0];
            if (!!wasPublic){
              database.ref(`isPublic/${publicPostId}`).remove();
            }
        });
      }
      database.ref(`users/${uid}/posts`).update({[id]: updates.isReadable});
      dispatch(editPost(id, updates));
    });
  };
};

// SET_POSTS
export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      if(!snapshot.val()) {
        let posts = [];
        dispatch(setPosts(posts));
      } else {
        return new Promise((resolve, reject) => {
          const postIds = [];
          snapshot.forEach((childSnapshot) => {
            postIds.push(childSnapshot.key);
          });
          let posts = [];
          postIds.forEach((postId, index)=>{
            database.ref(`posts/${postId}`).once('value').then((postSnapshot) => {
              posts.push({
                id: postSnapshot.key,
                ...postSnapshot.val()
              });
              if (index === postIds.length-1) {
                resolve(posts);
              }
            }).catch((error) => {
              console.log("error", error);
              reject(error);
            });
          });
        }).then((posts)=>{
          dispatch(setPosts(posts));
        }).catch(error => {
          console.log(error)
        });
      }
    }).catch((error)=>{
      console.log('error', error);
    });
  };
};
