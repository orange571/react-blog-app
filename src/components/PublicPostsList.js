import React from 'react';
import { connect } from 'react-redux';
import PublicPostListItem from './PublicPostListItem';

export const PublicPostsList = (props) => (
  <div className="content-container">
    <div className="content-container_public-posts">
      {
        props.posts.length === 0 ? (
          <div>
            <span>No posts</span>
          </div>
        ) : (
            props.posts.map((post) => {
              return <PublicPostListItem key={post.id} {...post} />;
            })
          )
      }
    </div>
  </div>
);

export default PublicPostsList;
