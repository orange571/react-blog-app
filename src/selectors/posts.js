import moment from 'moment';

// Get visible posts

export default (posts, { text, sortBy }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title.localeCompare(b.title);
    }
  });
};
