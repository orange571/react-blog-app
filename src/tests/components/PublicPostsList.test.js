import React from 'react';
import { shallow } from 'enzyme';
import PublicPostsList from '../../components/PublicPostsList';
import posts from '../fixtures/posts';

test('should render PublicPostList with posts', () => {
  const wrapper = shallow(<PublicPostsList posts={posts} />);
  expect(wrapper).toMatchSnapshot();
});
