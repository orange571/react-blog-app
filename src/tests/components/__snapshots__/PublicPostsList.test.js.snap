// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PublicPostList with posts 1`] = `
<div>
  <div
    className="content-container content-container--format-buttons show-for-desktop"
  >
    <button
      className="button--format false"
      onClick={[Function]}
    >
      <i
        className="fas fa-th"
      />
    </button>
    <button
      className="button--format active"
      onClick={[Function]}
    >
      <i
        className="fas fa-th-list"
      />
    </button>
  </div>
  <div
    className="content-container content-container--public-posts row"
  >
    <PublicPostListItem
      author="thisismytestname"
      body="Gum is rubbery"
      createdAt={0}
      id="1"
      isReadable={false}
      key="1"
      title="Gum"
      uid="thisismytestuid"
    />
    <PublicPostListItem
      author="thisismytestname"
      body="Rent is expensive"
      createdAt={-345600000}
      id="2"
      isReadable={true}
      key="2"
      title="Rent"
      uid="thisismytestuid"
    />
    <PublicPostListItem
      author="thisismytestname"
      body="Credit cards are thin"
      createdAt={345600000}
      id="3"
      isReadable={false}
      key="3"
      title="Credit Card"
      uid="thisismytestuid"
    />
  </div>
</div>
`;
